client/securedrop_client/gui/main.py

```diff
@@ -75,6 +80,13 @@ def __init__(
         self.main_pane.setLayout(layout)
         self.left_pane = LeftPane()
         self.main_view = MainView(self.main_pane, app_state)
+
+        # Other UI elements want to listen for UI updates, eg selections in the sourcelist.
+        # Provide a signal (request updates) and a slot (receive updates)
+        self.main_view.add_sourcelist_listener(
+            self.top_pane._source_requester(), self.top_pane._source_listener()
+        )
+
```

Comment:
There's a couple different ways we can do this, this one felt like the least-worst way, but I'm open to opinions.

Basically, the main view (holds the sourcelist) and the top bar (holds the toolbar) need a way of communicating; either the top bar receives a constant feed of events as the selected item of the sourcelist changes (in which case it would just connect to `onItemChanged` in the sourcelist and would not need to have a request signal, but it would need to push or pop items from the selected set all the time to keep track of the list); or, on demand it requests the selected set when it needs them (what I have implemented here).  I felt like this way was a little less error-prone.

I could have/could still skip the main_view and top_pane abstraction, and go for `self.main_view.source_list.connect....` etc, but that felt like reaching quite far into the business of another gui element; similarly,  `top_pane.batch_actions.toolbar.` could be accessed directly. I have no strong preferences and was trying make things fairly easy to read/follow. feedback welcome.

Reply:
How would you feel about expanding `SourceList.source_selection_changed` to take a `set()` of one or more sources (or IDs)?  That would give consistent signaling for this event and also align with the possible refactoring of things like `DeleteSourceJob(Source)` to `DeleteSourcesJob(set[Source])`.

https://github.com/freedomofpress/securedrop-client/blob/ddbb33aca725329309df098e828c2b9001c67760/client/securedrop_client/gui/widgets.py#L1097-L1100

Reply:
I don't know if it has any bearing on this decision but one issue we encounted 1.5 years ago (and haven't verified is still a thing recently) was this one https://github.com/freedomofpress/securedrop-client/issues/1567. Deleting multiple sources could be quite slow.  possibly could be relevant for deciding whether to store the state of 'selected sources' in one place or having the toolbar manage its own state. I haven't looked at the code properly recently enough to have any educated view on this though!

Reply:
@cfm: Looked at it, will look at it again, but iirc there're a few issues.
`source_selection_changed` informs which conversation is shown in the ConversationPane. (There will always be a need for 2 datapoints: currently-selected source aka last click in the sourcelist, if applicable; and "all selected sources", in multi-select context). 

I did think about refactoring source_selection_changed, source_selection_cleared, and multi_select to all be the same signal, which I think would be a better pattern, but I didn't want to scope creep too much and make the diff too big here, since that would have ripples through other parts of the sourcelist. 

I'm friendly to DeleteSourceJob refactoring, and there's no reason not to do it except for keeping the diff as small as possible; I'll go for it.

I had initially also refactored DeleteSourceAction along the same lines, but realized that wouldn't actually help us here, because while in the Source Menu the Action is created when the target Source is known, in the Toolbar, the Action is created _before_ the target source(s) are known since they can change dynamically.